lib/auth/init.go

```diff
@@ -322,59 +323,60 @@ func Init(cfg InitConfig, opts ...ServerOption) (*Server, error) {
 		log.Infof("Created default admin role: %q.", defaultRole.GetName())
 	}
 
-	// generate a user certificate authority if it doesn't exist
-	_, err = asrv.GetCertAuthority(types.CertAuthID{DomainName: cfg.ClusterName.GetClusterName(), Type: types.UserCA}, true)
-	if err != nil {
-		if !trace.IsNotFound(err) {
-			return nil, trace.Wrap(err)
-		}
-
-		log.Infof("First start: generating user certificate authority.")
-		userCA, err := generateSelfSignedCA(&cfg, asrv.keyStore, types.UserCA)
-		if err != nil {
-			return nil, trace.Wrap(err)
-		}
-		if err := asrv.Trust.UpsertCertAuthority(userCA); err != nil {
-			return nil, trace.Wrap(err)
-		}
-	}
-
-	// generate a host certificate authority if it doesn't exist
-	_, err = asrv.GetCertAuthority(types.CertAuthID{DomainName: cfg.ClusterName.GetClusterName(), Type: types.HostCA}, true)
-	if err != nil {
-		if !trace.IsNotFound(err) {
-			return nil, trace.Wrap(err)
-		}
-
-		log.Infof("First start: generating host certificate authority.")
-		hostCA, err := generateSelfSignedCA(&cfg, asrv.keyStore, types.HostCA)
+	// generate certificate authorities if they don't exist
+	for _, caType := range []types.CertAuthType{types.HostCA, types.UserCA, types.JWTSigner} {
+		caID := types.CertAuthID{Type: caType, DomainName: cfg.ClusterName.GetClusterName()}
+		ca, err := asrv.GetCertAuthority(caID, true)
 		if err != nil {
-			return nil, trace.Wrap(err)
-		}
-		if err := asrv.Trust.UpsertCertAuthority(hostCA); err != nil {
-			return nil, trace.Wrap(err)
+			if !trace.IsNotFound(err) {
+				return nil, trace.Wrap(err)
+			}
+			log.Infof("First start: generating %s certificate authority.", caID.Type)
+			if err := asrv.createSelfSignedCA(caID); err != nil {
+				return nil, trace.Wrap(err)
+			}
+		} else {
+			// Already have a CA. Make sure the keyStore has local keys.
+			if !asrv.keyStore.HasLocalActiveKeys(ca) {
+				// This could be one of a few cases:
+				// 1. A new auth server with an HSM being added to an HA cluster.
+				// 2. A new auth server with no HSM being added to an HA cluster
+				//    where all current auth servers have HSMs.
+				// 3. An existing auth server has restarted with a new HSM configured.
+				// 4. An existing HSM auth server has restarted no HSM configured.
+				// 5. An existing HSM auth server has restarted with a new UUID.
+				if ca.GetType() == types.HostCA {
```

Comment:
why not add an `AdditionalTrustedKey` for other CAs?

Reply:
It is not needed, we should not be signing anything with the User or JWT CA at this point right?

Reply:
Fair enough. Mention the fact that other CAs are not needed to issue the admin cert in the comment below.

Reply:
Fair enough. Mention the fact that other CAs are not needed to issue the admin cert in the comment below.

Reply:
done